### ISSUE TITLE
Problem with 'receiveDirect()' on SX1276

### ISSUE THREAD

**Issue by gef30441** [jgromes/RadioLib#276]
I tried to read the incoming data continuously via the receiveDirect() function on the following board: [Heltec WiFi LoRa ESP32 V2](https://heltec.org/project/wifi-lora-32/).
The board utilizes the SX1276, which is wired like this to the ESP32:
- CS: GPIO18
- DIO0: GPIO26
- DIO1: GPI35
- DIO2: GPI34

I basically used the 'SX127x FSK Modem Example' and extracted the code i figured was enough to fit my needs.
You can see my code here:
![code](https://user-images.githubusercontent.com/81627000/112988331-031e3300-9164-11eb-880d-d80dcb92eb48.PNG)

Unfortunately when flashing the software on my board, I get the following error message:
![image](https://user-images.githubusercontent.com/81627000/112988602-52fcfa00-9164-11eb-8590-f3eb7017ca9a.png)

Error code 16 translates to: [ERR_SPI_WRITE_FAILED](https://jgromes.github.io/RadioLib/group__status__codes.html#ga4e05af49324b9f065ae93f9664d51c2f).
The wiring can't be faulty (I hope at least) and I haven't found anything in the forums/RadioLib wiki that could help me.

I hope someone reading this is able to tell me what I did wrong here.

**Comment by gef30441** [comment]
I forgot to mention that I already looked at the signals on DIO2 (DATA) and DIO1 (CLK), but nothing meaningful was transferred.

**Comment by jgromes** [comment]
Double check the wiring, in your post it says that both DIO1 and DIO2 are connected to the same pin (GPI35), that seems wrong. Also, I suggest checking whether DIO2 is actually connected on that board, since on most LoRa/ESP32 boards, only DIO0 is always connected, and DIO1 is connected only sometimes. DIO2 is almost always unused.

Regarding your sketch, you're calling `receiveDirect` in the loop - you're basically spamming the radio with requests to set itself to Rx mode. Call `receiveDirect` once from setup, the radio will then remain in receive mode.

Finally, if neither of the above helps, please post debug mode output.

**Comment by gef30441** [comment]
I double checked the wiring and realized I made a typo.
DIO2 is on GPI34 and DIO1 is on GPI35. 

Thanks for your response, I will try your suggestion asap and report if it worked or not.

**Comment by gef30441** [comment]
So i just retried running the program with the following things changed:
- RadioLib_Debug active (RadioLib_Verbose inactive)
- receiveDirect() call in setup() instead of loop()

I get the following messages via Serial:
![image](https://user-images.githubusercontent.com/81627000/113258176-e43ab080-92cb-11eb-9e93-b95e361baa86.png)

I also noticed something strange while trying around (RadioLib_Debug still active): 
_Code:_
![image](https://user-images.githubusercontent.com/81627000/113262549-071b9380-92d1-11eb-9628-192e5895d9cb.png)

_Output:_
![image](https://user-images.githubusercontent.com/81627000/113262642-20bcdb00-92d1-11eb-893e-f78e5febbd76.png)

Basically, I initialized the SX1276 with default settings and then started directReceive(). As you can see in the output, it didn't work (DCLK doesn't change periodically) - even though there was no error message.

**Comment by jgromes** [comment]
The fact that it doesn't return the SPI error with default settings is interesting, I'll look into that.

You can't use Serial.prints to show the data and clock states, Serial takes way too long and the duration isn't always the same. You have to sample the data line based on an interrupt, and save them to a buffer, then print that. I'm currently working on implementing this sort of direct mode support.

**Comment by gef30441** [comment]
I made another interesting discovery: I initialized the SX1276 with default settings, except for the frequency of 868.0 MHz. 
When calling the receiveDirect() once I get the error 16 (as stated above). But when calling it another time (directly after the first time in setup()) there's no error. I think register 0x1 still has its old contents saved to itself when checking the first time.

But calling receiveDirect() twice doesn't always work. For some specific parameters (in my case I changed the bitrate to 300kbps and the RxBandwidth to 3.1kHz via their specific function setBitRate() and setRxBandwidth()) error 16 remains.

**Comment by gef30441** [comment]
I also tried printing the continuously received data.

![image](https://user-images.githubusercontent.com/81627000/114015854-7b72ab80-986a-11eb-8745-17fcf3b3687f.png)
![image](https://user-images.githubusercontent.com/81627000/114015910-8c232180-986a-11eb-885d-8074ef1bfe2f.png)
![image](https://user-images.githubusercontent.com/81627000/114015950-9c3b0100-986a-11eb-8848-f4ad2a566862.png)

Line 15 - 24: Naming the respective pins.
Line 28 - 43: Implementing the interrupt handler.
Line 47 - 48: I don't know if it's necessary to declare the pinModes explicitly here.
Line 51 : Attach interrupt to pin 35 at rising edge (at rising edge reading data is possible, at falling edge data is refreshed, more info in the [SX1276 datasheet](https://www.semtech.com/products/wireless-rf/lora-transceivers/sx1276).

How this basically works: In DCLK_interrupt() all contents in testvar are shifted a bit to the left. The current state of the data pin is saved to the "new" free bit on the right.
When the testvar is full (64 bits are filled in from data), the counter reached 64 and the receiveFull flag is set true. The counter gets a reset and the contents of testvar are deleted after getting copied to buffer.

The loop() constantly checks if the receiveFull flag is set. If yes, the buffer is printed via Serial.print and the receiveFull flag is reset.

**Comment by jgromes** [comment]
> I initialized the SX1276 with default settings, except for the frequency of 868.0 MHz.
When calling the receiveDirect() once I get the error 16 (as stated above). But when calling it another time (directly after the first time in setup()) there's no error. I think register 0x1 still has its old contents saved to itself when checking the first time.

That's possible - I did notice already that some registers take time to update, especially the OpMode register. Though it seems strange it would depend on the value of the parameter.

Regarding the direct DIO sampling - that's very similar to what @yeckel has done with his direct mode D-Star receiver: https://github.com/yeckel/D-StarReceive

I'm also working on RadioLib implementation of this direct mode approach that would be available to all modules that can do this, not just SX127x.

**Comment by yeckel** [comment]
Hi @gef30441 I've observed the same issue. When starting AFC I could not call ```setMode(SX127X_RX)``` however it seems to me that the RX mode **is actually set!** Just when the register is read back it contains the old bit value. Even after looong timeout. Probably you could just ignore that error :-D  Surprisingly ```setMode(SX127X_TX)``` works fine (changes the same bit).

**Comment by jgromes** [comment]
I wasn't able to find anything in datasheet or errata that could explain this behavior, so let's just patch it in RadioLib - checking of that one bit was disabled in the latest commit, only for FSK RX.

@gef30441 @yeckel could you guys give it a try now?

**Comment by yeckel** [comment]
Yes, it looks fine and works. I've tested it with my AFC. Good idea.

**Comment by jgromes** [comment]
Thanks, looks like this has been resolved then - feel free to reopen if it hasn't.
